### PR TITLE
add transform for uniform array move out

### LINKED ISSUE #34708: [MIR] double drop with slice patterns
## Version

```
$ rustc -V
rustc 1.11.0-dev (42b7c32ac 2016-07-02)
```
## STR

``` Rust
#![feature(slice_patterns, rustc_attrs)]

struct NoisyDrop;
impl Drop for NoisyDrop {
    fn drop(&mut self) { println!("splat!"); }
}

#[rustc_mir]
fn main() {
    let [x] = [NoisyDrop];
}
```
## Expected Result

`NoisyDrop` gets dropped once
## Actual Result

`NoisyDrop` is dropped twice

### PR DESCRIPTION
**Pull request by mikhail-m1**
reworked second step for fix #34708
previous try #46686
r? @nikomatsakis

### CODE REVIEW

**Comment by arielb1** [comment]
This looks like the KISS approach that might just enable us to stabilize slice patterns.

File: src/librustc_mir/transform/uniform_array_move_out.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This pass converts move out from array by Subslice and
+// ConstIndex{.., from_end: true} to ConstIndex move out(s) from begin
+// of array. It allows detect error by mir borrowck and elaborate
+// drops for array without additional work.
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/uniform_array_move_out.rs]: I think that, to document this sort of thing, it is best if we can start with a comment that includes some kind of "before/after" -- e.g., show the MIR that we are starting with, and the MIR that we are ending with. Then, in the actual code, we can leave comments that reference this running example.

File: src/librustc_mir/transform/uniform_array_move_out.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                let opt_size = size_o.val.to_const_int().and_then(|v| v.to_u64());
+                if let Some(size) = opt_size {
+                    let size = size as u32;
+                    match proj.elem {
+                        ProjectionElem::Subslice{from, to} => {
+                            patch.make_nop(location);
+                            let to = size - to;
+                            let mut temps = vec![];
+                            for _i in from..to {
+                                let _temp = patch.new_temp(_t, mir.source_info(location).span);
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/uniform_array_move_out.rs]: Nit: there should never be a use of a variable named with `_t` -- the leading `_` is a signal that you *don't* use the variable... can we give it some other name? (Same with `_temp`)

File: src/librustc_mir/transform/uniform_array_move_out.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            let place_ty = proj.base.ty(mir, tcx).to_ty(tcx);
+            if let ty::TyArray(_t, size_o) = place_ty.sty {
+                let opt_size = size_o.val.to_const_int().and_then(|v| v.to_u64());
+                if let Some(size) = opt_size {
+                    let size = size as u32;
+                    match proj.elem {
+                        ProjectionElem::Subslice{from, to} => {
+                            patch.make_nop(location);
+                            let to = size - to;
+                            let mut temps = vec![];
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/uniform_array_move_out.rs]: Nit: doesn't matter, but I would personally write this with an iterator

```
let temps: Vec<_> =
    (from..to)
    .map(|i| {
       let temp = patch.new_temp(...);
       ...
       temp
    })
    .collect();
```

File: src/librustc_mir/transform/uniform_array_move_out.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                                let _temp = patch.new_temp(_t, mir.source_info(location).span);
+                                patch.add_statement( location, StatementKind::StorageLive(_temp));
+                                patch.add_assign(location,
+                                                 Place::Local(_temp),
+                                                 Rvalue::Use(
+                                                     Operand::Move(
+                                                         Place::Projection(box PlaceProjection{
+                                                             base: proj.base.clone(),
+                                                             elem: ProjectionElem::ConstantIndex{
+                                                                 offset: _i,
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/uniform_array_move_out.rs]: Nit: Name this `i`, not `_i`

File: src/librustc_mir/transform/uniform_array_move_out.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                                     src_place: &Place<'tcx>,
+                                     patch: &mut MirPatch<'tcx>) {
+        if let Place::Projection(ref proj) = *src_place {
+            let place_ty = proj.base.ty(mir, tcx).to_ty(tcx);
+            if let ty::TyArray(_t, size_o) = place_ty.sty {
+                let opt_size = size_o.val.to_const_int().and_then(|v| v.to_u64());
+                if let Some(size) = opt_size {
+                    let size = size as u32;
+                    match proj.elem {
+                        ProjectionElem::Subslice{from, to} => {
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/uniform_array_move_out.rs]: Example comment:

---

Before we have:

```
X = move Y[3..5]
```

we want to transform to the following patch:

```
- X = move Y[3..5]
+ tmp0 = move Y[3];
+ tmp1 = move Y[4];
+ X = [tmp0, tmp1];
```

File: src/librustc_mir/transform/uniform_array_move_out.rs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                                             dst_place.clone(),
+                                             Rvalue::Aggregate(box AggregateKind::Array(_t),
+                                             temps.iter().map(
+                                                 |x| Operand::Move(Place::Local(*x))).collect()
+                                             ));
+                            for temp in temps {
+                                patch.add_statement(location, StatementKind::StorageDead(temp));
+                            }
+                        }
+                        ProjectionElem::ConstantIndex{offset, min_length: _, from_end: true} => {
```
Comment (nikomatsakis) [inline src/librustc_mir/transform/uniform_array_move_out.rs]: Some kind of comment would be good here. Probably:

---

Normalize "from-end" indices to operate from the start.

File: src/test/compile-fail/borrowck/borrowck-move-out-from-array.rs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    let [.., _y] = a; //[ast]~ ERROR [E0382]
+                      //[mir]~^ ERROR [E0382]
+}
+
+fn move_out_by_const_index_and_subslice() {
+    let a = [box 1, box 2];
+    let [_x, _] = a;
+    let [_y..] = a; //[ast]~ ERROR [E0382]
+                    //[mir]~^ ERROR [E0382]
+}
```
Comment (nikomatsakis) [inline src/test/compile-fail/borrowck/borrowck-move-out-from-array.rs]: I think this transformation should have a `src/test/mir-opt` test that shows the before/after for various interesting cases. Can you add that?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Very nice! Left various style nits and also some testing suggestions.

**Comment by pietroalbini** [comment]
@nikomatsakis the author seems to have fixed your concerns, could you review this again?

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors r+

Very cool

**Pull request merged**